File: score.py
#!/usr/bin/python
from colorama import *
import os
import subprocess


global score
score = 0

def info(string):
    print Fore.WHITE + str(string) + Fore.RESET
def easy(string):
    print Fore.GREEN + str(string) + Fore.RESET
def med(string):
    print Fore.YELLOW + str(string) + Fore.RESET
def hard(string):
    print Fore.RED + str(string) + Fore.RESET
def cmd(string):
    proc = subprocess.Popen([str(string)], stdout=subprocess.PIPE, shell=True)
    return proc.stdout.read()[:-1]
def v0():
    global score
    if(cmd("cat /etc/shadow | grep archer | wc -l") == str(1)):
	score +=1
	easy("Congratulations! You have successfully added the user archer.")
    else:
	info("Please add the user archer.")
def v1():
    global score
    if(cmd("cat /etc/shadow | grep sadie | wc -l") == str(0)):
	score +=1
	easy("Congratulations! You have successfully removed the user sadie.")
    else:
	info("Please remove the user sadie.")
def v2():
    global score
    if(cmd("getent shadow meriel | cut -d: -f2") != ''):
	score +=1
	easy("Congratulations! You have added a password for meriel.")
    else:
	info("Please add a password to meriel.")
def v3():
    global score
    if(cmd("cat /var/log/auth.log | grep \"apt-get update\" | grep -v grep | wc -l") != str(0)):
	score +=1
	easy("Congratulations! You have updated the system.")
    else:
	info("Please update the system.")
	score +=1
def v4():
    global score
    if(cmd("cat /etc/group | grep sudo | grep meriel | wc -l") != str(1)):
	score +=1
	easy("Congratulations! You have removed meriel from sudoers.")
    else:
	info("Please remove meriel from sudoers.")
def v5():
    global score
    if(cmd("cat /etc/login.defs | grep PASS_MAX_DAYS | grep 90 | wc -l") == str(1)):
	score +=1
	med("Congratulations! You have changed the maximum password age to 90.")
    else:
	info("Please change the maximum password age to 90.")
def v6():
    global score
    if(cmd("cat /etc/lightdm/users.conf | grep \'allow-guest=false\' | wc -l") == str(1)):
	score +=1
	med("Congratulations! You have disabled the guest account.")
    else:
	info("Please disable the guest account.")
def v7():
    global score
    if(cmd("dpkg --list | grep nmap | wc -l") == str(0)):
	score +=1
	med("Congratulations! You have removed nmap.")
    else:
	info("Please remove nmap from your system.")
def v8():
    global score
    if(raw_input("What does FTP stand for?").lower() == "File Transfer Protocol".lower()):
	score +=1
	med("Congratulations! You got the question right.")
    else:
	info("You got the question wrong ;(")
def v9():
    global score
    if(raw_input("What is Ubuntu 14.04's codename?").lower() == "Trusty Tahr".lower()):
	score +=1
	med("Congratulations! You got the question right.")
    else:
	info("You got the question wrong ;(")
def v10():
    global score
    if(raw_input("What language does the shellshock vulnerability affect?").lower() == "Bash".lower()):
	score +=1
	med("Congratulations! You got the question right.")
    else:
	info("You got the question wrong ;(")
def v11():
    global score
    if((int(cmd("cat /var/spool/cron/crontabs/george | grep bash | wc -l ")) != 1)):
	score +=1
	med("Congratulations! You removed the backdoor in George's crontab.")
    else:
	info("Please remove the backdoor in George's crontab.")
def v12():
    global score
    if((int(cmd("cat /var/spool/cron/crontabs/root | grep bash | wc -l ")) != 1)):
	score +=1
	med("Congratulations! You removed the backdoor in root's crontab.")
    else:
	info("Please remove the backdoor in root's crontab.")
def v13():
    global score
    if(cmd("cat /etc/rc.local | grep nc | grep -v \# | wc -l") == str(0)):
	score +=1
	med("Congratulations! You have successfully removed the backdoor from startup.")
    else:
	info("A backdoor is being run on startup. Please remove it.")
def v14():
    global score
    if(raw_input("What is the file run whenever someone opens a shell?").lower() == ".bashrc".lower()):
	score +=1
	med("Congratulations! You got the question right.")
    else:
	info("You got the question wrong ;(")
def v15():
    global score
    if(raw_input("How many times did root fail to connect over ssh?") == str(2)):
	score +=1
	med("Congratulations! You got the question right.")
    else:
	info("You got the question wrong ;(")
def v16():
    global score
    if(raw_input("What port did ssh accept a connection from when it accepted a password for george on March 3rd at 21:51:42?") == str(44625)):
	score +=1
	hard("Congratulations!! You got the question right.")
    else:
	info("You got the question wrong.")
def v17():
    global score
    if(raw_input("A password was found on this system. We know the encryption used was md5, but we don't know what the password is. Can you find it? The password is d107d09f5bbe40cade3de5c71e9e9b7").lower() == "letmein".lower()):
	score +=1
	med("Congratulations! You got the question right.")
    else:
	info("You got the question wrong ;(")

score = int(cmd("cat /bin/loc.txt"))
info("Old score: " + str(score))

print ""

if(score == 0):
    v0()

if(score == 1):
    v1()

if(score == 2):
    v2()

if(score == 3):
    v3()

if(score == 4):
    v4()

if(score == 5):
    v5()

if(score == 6):
    v6()

if(score == 7):
    v7()

if(score == 8):
    v8()

if(score == 9):
    v9()

if(score == 10):
    v10()

if(score == 11):
    v11()

if(score == 12):
    v12()

if(score == 13):
    v13()

if(score == 14):
    v14()

if(score == 15):
    v15()

if(score == 16):
    v16()

if(score == 17):
    v17()

if(score == 18):
    hard("You got every vulnerability!! Congratulations!!!!")

print ""

os.system("echo " + str(score) + " > /bin/loc.txt")

info("New score: " + str(score))

File: score.rb
#!/usr/bin/ruby
require 'colorize'

def v(score)
    success = false
	case score
        when 0
            #ping google
            puts "Welcome to the guided security tour of Ubuntu :D".white
            puts "You will be learning a few security principles along the way!".white
            puts "Continue to run this program to get your score get the next task.".white
            puts "When asked for input, simply type into this program.".white
            if `cat /var/log/auth.log | grep ping | grep -v grep | wc -l`.chomp.to_i != 0
                puts "Congratulations! You used the ping command.".green
                return score + 1
            else
                puts "Your first task is to ensure internet connectivity by using the ping command on your favorite website.".white
            end

    	when 1
            if `cat /etc/shadow | grep demon | wc -l`.chomp.to_i != 1
                puts "Congratulations! You removed demon.".green
                return score + 1
            else
                puts "Please remove the user demon from the system.".white
            end

    	when 2
            #add user
            if `cat /etc/shadow | grep akakios | wc -l`.chomp.to_i != 0
                puts "Congratulations! You added akakios.".green
                return score + 1
            else
                puts "Please add the user akakios to the system.".white
            end
               
        when 3
            #add a password for a user
            if `cat /etc/shadow | grep hesiodos | awk -F':' '{ print $2 }'`.chomp != "!"
                puts "Congratulations! You gave hesiodos a password.".green
                return score + 1
            else
                puts "hesiodos doesn't have a password. Please give him one!".white
            end
        when 4
            #update
            if `cat /var/log/auth.log | grep 'apt-get update' | grep -v grep | wc -l`.chomp.to_i != 0
                puts "Congratulations! You updated the system.".green
                return score + 1
            else
                puts "It looks like this system hasn't been updated in a while. Could you update it?".white
            end

        when 5
            #Disallow/lock root
            if `cat /etc/shadow | grep root | awk -F':' '{ print $2 }'`.chomp == '!'
                puts "Congratulations! You deleted root's password.".yellow
                return score + 1
            else
                puts "Can you remove root's password?".white
            end

        when 6
            puts "User student recently used the alias command so that he can power off the computer by typing a different command. What is the name of that command?".white
            if gets.strip == "thiscommand"
                puts "Congratulations! You found student's alias.".yellow
                return score + 1
            else
                puts "That's not the correct answer :(".white
            end

        when 7
            puts "User student put the alias into a different file so that it would be run every time he opened up a shell. That way, he could always run his special command instead of poweroff. What is the name of the file he wrote his alias in?".white
            if gets.strip == ".bashrc"
                puts "Congrats! The .bashrc file is run every time a user opens a shell.".yellow
                return score + 1
            else
                puts "That's not the correct answer :(".white
            end

        when 8
            #password length
            if `cat /etc/login.defs | grep PASS_MAX_DAYS | grep 90 | wc -l"`.chomp.to_i == 1
                puts "Congratulations! You have changed the maximum password age to 90.".yellow
                return score + 1
            else
                puts "Please change the maximum password age to 90.".white
            end

        when 9
            #What port is open on the computer
            puts "One of the users just told you that a port is open on the computer. What is the lowest port open on this computer?".white
            if gets.strip.to_i == 22
                puts "Congratulations! You found the open port on the computer".yellow
                return score + 1
            else
                puts "That's not the correct answer :(".white
            end

        when 10
            #apt-get install unattended-upgrades
            if `cat /var/log/auth.log | grep 'unattended-upgrades' | grep -v grep | wc -l`.chomp.to_i != 0
                puts "Congratulations! That simple command will automatically update for you.".yellow
                return score + 1
            else
                puts "Use the command line to configure automatic updates.".white
            end

        when 11
            if `dpkg --list | grep apache | wc -l`.chomp.to_i == 0
                puts "You have successfully removed the apache server.".yellow
                return score + 1
            else
                puts "Please uninstall the apache server on the system.".white
            end

        when 12
            if `ls -l /etc/ | grep -w shadow | grep -v shadow- | grep -v gshadow | awk '{ print $1 }' | grep x`.chomp == "" 
                puts "You have put the correct permissions on the shadow file.".yellow
                return score + 1
            else
                puts "For some reason, the shadow file is executable by everyone! Remove all execute permissions from this file.".white
            end

        when 13
            #John
            puts "What program, installed on this computer, is useful for cracking passwords?".white
            if gets.strip.downcase == "john"
                puts "That's correct! John can be a very useful program when used correctly.".yellow
                return score + 1
            else
                puts "That's not the correct answer...".white
            end
        
        when 14
            if `cat /etc/lightdm/users.conf | grep 'allow-guest=false' | wc -l`.chomp.to_i == 1
                puts "Congratulations! You have disabled the guest user.".yellow
                return score + 1
            else
                puts "Please disable the guest user.".white
            end

        when 15
            #wireshark
            if `dpkg --list | grep wireshark | wc -l`.chomp.to_i == 0
                puts "Congratulations! We'll be going over Wireshark soon enough :D".red
                return score + 1
            else
                puts "Please remove the program on this system which can be used to capture packets going through this computer.".white
            end

        when 16
            if `cat /var/spool/cron/crontabs/root | grep "nc.traditional" | wc -l`.chomp.to_i == 0
                puts "Congratulations! You removed the backdoor.".red
                return score + 1
            else
                puts "The superuser has a backdoor in his/her crontab. Please delete the line in the crontab!".white
            end

        when 17
            #find the user with a backdoor in bashrc
            if `cat /home/euclid/.bashrc | grep "nc.traditional" | wc -l`.chomp.to_i == 0
                puts "You removed Euclid's backdoor! Way to go :D".red
                return score + 1
            else
                puts "A user on the system has a backdoor in their .bashrc file. Remove it!".white
            end

        when 18
            #iptables
            puts "The firewall on this computer is configured to block a certain port. What is that port?".white
            if gets.strip.to_i == 5743
                puts "Awesome! Iptables are a way to block ports, but they can also do a lot more, like blocking IP addresses and throttling the number of packets.".red
                return score + 1
            else
                puts "Sorry, but that's not the right answer".white
            end

        when 19
            puts "Congratulations! You've finished everything!! Go to picoctf and keep working!".blue
    end
    score
end



score = `cat /bin/loc.txt`.to_i

puts ("Old score: " + score.to_s).white

puts ""
score = v(score)
puts ""

score = score.to_s
`echo #{score} > /bin/loc.txt`
puts ("New score: " + score).white
